File: LC2130-MaximumTwinSumOfALinkedList.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
int pairSum(struct ListNode* head) {
    /* Solution keypoints:
     *     1. Find the middle node - the starting point of the 2nd half.
     *     2. Reverse the 2nd half linked list
     *     3. Get the maximum twin sum with two pointers moving forward
     */

    /* 1. Find the middle node with fast and slow pointers */
    struct ListNode *slow = head;
    struct ListNode *fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    /* 2. Reverse the 2nd half linked list */
    struct ListNode* prev = NULL;
    struct ListNode* curr = slow;  // start from the middle node
    struct ListNode* nextNode;

    while (curr) {
        nextNode = curr->next;
        curr->next = prev;
        prev = curr;
        curr = nextNode;
    }

    /* 3. Get the maximum twin sum with two pointers moving forward */
    int maxSum = 0;
    struct ListNode* p1 = head;
    struct ListNode* p2 = prev;
    while (p2) {
        int sum = p1->val + p2->val;
        if (sum > maxSum) maxSum = sum;
        p1 = p1->next;
        p2 = p2->next;
    }

    return maxSum;
}


// Time complexity: O(n) - we traverse the linked list a constant number of times
// Space complexity: O(1) - we use a constant amount of extra space

File: LC2130-MaximumTwinSumOfALinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        # Step 1: Find the middle of the linked list using slow and fast pointers
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Step 2: Reverse the second half of the linked list
        prev = None
        curr = slow

        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        # Step 3: Calculate the maximum twin sum
        max_twin_sum = 0
        first_half = head
        second_half = prev

        while second_half:
            twin_sum = first_half.val + second_half.val
            max_twin_sum = max(max_twin_sum, twin_sum)
            first_half = first_half.next
            second_half = second_half.next

        return max_twin_sum